src/libAtomVM/otp_socket.c

```diff
@@ -0,0 +1,1364 @@
+            }
+        }
+
+        int res = close(fd_obj->fd);
+        if (UNLIKELY(res != 0)) {
+            AVM_LOGW(TAG, "Failed to close socket %i", res);
+        }
+
+        TRACE("nif_socket_close: Clearing pid for socket fd=%i\n", fd_obj->fd);
+        fd_obj->fd = 0;
+        fd_obj->selecting_process_id = INVALID_PROCESS_ID;
+        fd_obj->pid = term_invalid_term();
+        fd_obj->ref_ticks = 0;
+    } else {
+        TRACE("Double close on socket fd %i", fd_obj->fd);
+    }
+
+    return OK_ATOM;
+}
+
+//
+// select
+//
+
+static term nif_socket_select(Context *ctx, term argv[], enum ErlNifSelectFlags mode)
+{
+    TRACE("nif_socket_select\n");
+
+    VALIDATE_VALUE(argv[0], term_is_socket);
+
+    term process_pid_term = argv[1];
+    VALIDATE_VALUE(process_pid_term, term_is_pid);
+    int32_t process_pid = term_to_local_process_id(process_pid_term);
+    TRACE("process_pid=%i\n", (int) process_pid);
+
+    term select_ref_term = argv[2];
+    if (select_ref_term != UNDEFINED_ATOM) {
+        VALIDATE_VALUE(select_ref_term, term_is_reference);
+    }
+    void *fd_obj_ptr;
+    if (UNLIKELY(!enif_get_resource(erl_nif_env_from_context(ctx), get_socket(argv[0]), socket_resource_type, &fd_obj_ptr))) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct SocketFd *fd_obj = (struct SocketFd *) fd_obj_ptr;
+    TRACE("fd_obj->fd=%i\n", (int) fd_obj->fd);
+
+    ErlNifEnv *env = erl_nif_env_from_context(ctx);
+    if (fd_obj->selecting_process_id != process_pid && fd_obj->selecting_process_id != INVALID_PROCESS_ID) {
+        if (UNLIKELY(enif_demonitor_process(env, fd_obj, &fd_obj->selecting_process_monitor) != 0)) {
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        fd_obj->selecting_process_id = INVALID_PROCESS_ID;
+    }
+    // Monitor first as select is less likely to fail and it's less expensive to demonitor
+    // if select fails than to stop select if monitor fails
+    if (fd_obj->selecting_process_id != process_pid) {
+        if (UNLIKELY(enif_monitor_process(env, fd_obj, &process_pid, &fd_obj->selecting_process_monitor) != 0)) {
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        fd_obj->selecting_process_id = process_pid;
+    }
+
+    if (UNLIKELY(enif_select(erl_nif_env_from_context(ctx), fd_obj->fd, mode, fd_obj, &process_pid, select_ref_term) < 0)) {
+        enif_demonitor_process(env, fd_obj, &fd_obj->selecting_process_monitor);
+        fd_obj->selecting_process_id = INVALID_PROCESS_ID;
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    TRACE("nif_socket_select: Setting pid for socket fd %i to %i\n", fd_obj->fd, process_pid);
+    fd_obj->pid = process_pid_term;
+    fd_obj->ref_ticks = (select_ref_term == UNDEFINED_ATOM) ? 0 : term_to_ref_ticks(select_ref_term);
+
+    return OK_ATOM;
+}
+
+static term nif_socket_select_read(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    return nif_socket_select(ctx, argv, ERL_NIF_SELECT_READ);
+}
+
```

Comment:
Re-reading OTP's documentation on socket:send, I confirm we need (a) to configure socket to be non blocking when they are open, (b) to implement select_write and (c) have a small chunk of Erlang code that loops to select_write and send until an error occurs or all data is sent.

Erlang/OTP does it in C, but I believe Erlang would do:

https://github.com/erlang/otp/blob/b5e045c8c3bbeb6ad6f66e38cf7140c045890d00/erts/emulator/nifs/unix/unix_socket_syncio.c#L5172

Reply:
Are you okay with a blocking implementation for now?